File: leetcode/Swift/No20.valid-parentheses.swift
/*
 * Difficulty:
 * Easy
 *
 * Desc:
 * Given a string containing just the characters '(', ')', '{', '}', '[' and ']',
 * determine if the input string is valid.
 * The brackets must close in the correct order, "()" and "()[]{}" are all valid but "(]" and "([)]" are not.
 *
 * 给出一个含有括号的字符串，判断其是否正确的闭合
 * 例如，'[](){}' 是正确闭合的，'{[()]}' 也是正确闭合的，但 '[](}' 这样的就不是
 */

class Solution {
    let charMap: [Character:Character] = [
        "}": "{",
        ")": "(",
        "]": "["
    ]

    func isValid(_ s: String) -> Bool {
        var stack = [Character]()
        for char in s {
            let c = charMap[char]
            if c == nil {
                stack.append(char)
            } else {
                if stack.popLast() != c {
                    return false
                }
            }
        }
        return stack.count == 0
    }
}

File: leetcode/JavaScript/No20.valid-parentheses.js
/*
 * Difficulty:
 * Easy
 *
 * Desc:
 * Given a string containing just the characters '(', ')', '{', '}', '[' and ']',
 * determine if the input string is valid.
 * The brackets must close in the correct order, "()" and "()[]{}" are all valid but "(]" and "([)]" are not.
 *
 * 给出一个含有括号的字符串，判断其是否正确的闭合
 * 例如，'[](){}' 是正确闭合的，'{[()]}' 也是正确闭合的，但 '[](}' 这样的就不是
 */

/**
* @param {string} s
* @return {boolean}
*
* 思路：运用栈
*/
var isValid = function(s) {
  const stack = []
  const BracketsMap = {
    ')': '(',
    ']': '[',
    '}': '{'
  }

  for (var i = 0; i < s.length; i += 1) {
    var str = s[i]
    if (BracketsMap[str]) {
      if (stack.pop() !== BracketsMap[str]) {
        return false
      }
    } else {
      stack.push(str)
    }
  }
  return stack.length === 0
}
